[WebDriver][win] WebDriver launches MiniBrowser for automation

haruhisa-shin opened the pull request:
#### 7e00247af20ab27c5d35dc61e7372cc0edbb9175
<pre>
[WebDriver][win] WebDriver launches MiniBrowser for automation
<a href="https://bugs.webkit.org/show_bug.cgi?id=282972">https://bugs.webkit.org/show_bug.cgi?id=282972</a>

Reviewed by Fujii Hironori.

The win port WebDriver supports connecting to a browser
that is already running.
However, win port WebDriver has not yet implemented
the feature of launching the browser by itself.
This makes it difficult to run the WebDriverTests.

This change adds a feature to the WebDriver to launch
a browser to connect if target is not specified.

Before this change, the --target(-t) option is
mandatory for win port WebDriver.

This change allows the --target option to be omitted.
In that case, the WebDriver will scan freeport and
launching a MiniBrowser that listens on 127.0.0.1:(freeport).
This MiniBrowser is hidden, but WebDriver can connect to
and automate it.

Canonical link: <a href="https://commits.webkit.org/287053@main">https://commits.webkit.org/287053@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/ced1da5d486a34ad09bbfc9c2fa0c374f0c7bdb6

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/78136 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/57180 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/31512 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/82794 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/29398 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/66330 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/5464 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/61187 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/19105 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/81203 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/51176 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/67965 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/41501 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/48535 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/24733 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/27736 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/69630 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/25089 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/84157 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/5502 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/122/builds/3720 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/69409 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/5659 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/67083 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/68664 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/17130 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/12664 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/10904 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/5451 "Built successfully") | [  ~~🛠 mac-safer-cpp~~](https://ews-build.webkit.org/#/builders/120/builds/8203 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/5440 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/8872 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/7229 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

fujii commented on Source/WebDriver/SessionHost.cpp:
```diff
@@ -105,7 +105,7 @@ void SessionHost::dispatchMessage(const String& message)
 }
 
-#if !USE(GLIB)
+#if !USE(GLIB) && !USE(INSPECTOR_SOCKET_SERVER)
```
Which build configuration does match with `!USE(GLIB) && !USE(INSPECTOR_SOCKET_SERVER)`? I think this is dead code after this change. Right?

haruhisa-shin replied:
I'm not familiar with other ports, but for example, wouldn't the iOS fall under that criteria?
I thought there were many build configurations that did not use either GLIB or INSPECTOR_SERVER_SOCKET.

fujii replied:
Apple doesn't use this WebDriver implementation.

haruhisa-shin replied:
I found 4 *.cmake files in the Source/WebDriver folder, Platform(GTK|PlayStation|Win|WPE).cmake.
Is it correct to assume that these are the only 4 types of build configurations now?
If that is correct, I can revise the conditions.

fujii replied:
Apple doesn't use CMake. The number of .cmake files doesn't match with the number of build configurations.

haruhisa-shin replied:
I see, then, I'll change the function `SessionHost::isRemoteBrowser()` to common code.
It will return false by default and true only for remote automation.

fujii commented on Source/WebDriver/socket/SessionHostSocket.cpp:
```diff
@@ -212,4 +330,11 @@ void SessionHost::sendMessageToBackend(const String& message)
 }
 
+#if USE(INSPECTOR_SOCKET_SERVER)
+bool SessionHost::isRemoteBrowser() const
```
This looks same with SessionHostGlib.cpp. Why don't you unify them?

haruhisa-shin replied:
I'll unify them and move it to SessionHost.cpp.

fujii commented on Source/WebDriver/socket/SessionHostSocket.cpp:
```diff
@@ -212,4 +330,11 @@ void SessionHost::sendMessageToBackend(const String& message)
     sendWebInspectorEvent(sendEvent->toJSONString());
 }
 
+#if USE(INSPECTOR_SOCKET_SERVER)
```
Do you need `USE(INSPECTOR_SOCKET_SERVER)` in this file?

fujii commented on Source/WebDriver/SessionHost.h:
```diff
@@ -145,6 +145,10 @@ class SessionHost final
 #elif USE(INSPECTOR_SOCKET_SERVER)
     Function<void(bool, std::optional<String>)> m_startSessionCompletionHandler;
     std::optional<Inspector::ConnectionID> m_clientID;
+    bool m_isRemoteBrowser { true };
```
I think we can share above m_isRemoteBrowser with USE(GLIB).

haruhisa-shin replied:
Currently `USE(INSPECTOR_SOCKET_SERVER)` is enabled in win and playstation ports.
Either is fine for win, but I think true is better for playstation.

fujii replied:
I don't understand. Why problem happens for PlayStation port if the initial value is `true`. I think `SessionHost::connectToBrowser` set the variable `false`.

haruhisa-shin replied:
> Why problem happens for PlayStation port if the initial value is true.

No problems have occurred if the initial value is true. I said "I think true is better for playstation."
This is because PlayStation's WebDriver only supports remote automation.

Therefore, I think that the initial value as socket port should be `true`, and only `false` if the browser is launched in Windows.
What do you think?

haruhisa-shin replied:
^ Sorry. Please ignore the above comment.
The `SessionHost::isRemoteBrowser()` commonality will fix this part too.

fujii commented on Source/WebDriver/socket/SessionHostSocket.cpp:
```diff
@@ -69,9 +154,35 @@ void SessionHost::connectToBrowser(Function<void (std::optional<String> error)>&
+        StringBuilder envVarBuilder;
+        envVarBuilder.append(targetIp);
+        envVarBuilder.append(":"_s);
+        envVarBuilder.append(targetPort);
```
Use makeString.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.h:
```diff
@@ -40,14 +42,21 @@ namespace WebKit {
     String m_sessionIdentifier;
+
+    static void close(WKPageRef, const void*);
+    HashMap<HWND, RefPtr<WebView>> m_webViews;
```
Why does Windows port have to remember WebView instances? Other ports don't do that. The HWND will leak?

haruhisa-shin replied:
Something must have gone wrong, but I forgot...
I'll try to disable this code and see what happens.

haruhisa-shin replied:
The reason is that the window instance created by `WebView::create()` must be retained or it will exit immediately.
FYI, the inspector's subwindow is retained as `m_webView` for the same reason.
https://github.com/WebKit/WebKit/blob/main/Source/WebKit/UIProcess/Inspector/win/RemoteWebInspectorUIProxyWin.cpp#L131

Multiple windows could be created here and all of them needed to be retained.
So I created `retainWebView()` function.

fujii replied:
I think Windows port is doing something wrong with lifetime of these objects because other ports don't do that.

haruhisa-shin replied:
Maybe so, but I would like to fix it separately.

fujii replied:
We no longer have the reason to use `HWND` here. We can use `WebPageProxy*` or `NativeWindowHandle`.
I think WebView should call ref() at startup and call deref() when the window is closed. 
If my understanding is correct, we should fix the lifetime problem before this change.

fujii replied:
I'm going to try that. Hold on please.

fujii replied:
I was wrong.
`requestNewPageWithOptions` calls `WebView::create` with null `parentWindow`.
It's a message-only window.
https://learn.microsoft.com/en-us/windows/win32/winmsg/window-features#message-only-windows
Then, the window of the view never gets closed.

fujii commented on Source/WebDriver/socket/SessionHostSocket.cpp:
```diff
@@ -30,10 +30,21 @@
 {
+#if PLATFORM(WIN)
+    if (m_browserHandle) {
+        ::TerminateProcess(*m_browserHandle, 0);
```
You need `CloseHandle` here. It's better to use `Win32Handle` for `m_browserHandle`.

haruhisa-shin replied:
I see. I'll try it.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -42,22 +44,43 @@ AutomationSessionClient::AutomationSessionClient(const String& sessionIdentifier
-    if (!found)
-        completionHandler(nullptr);
+    for (auto webView : m_webViews)
+        ::SendMessage(webView.key, WM_DESTROY, 0, 0);
```
It's strange that we have to send native message here even though we have `WebView` object.
I think we should call `webView.close()` or just destroy the object becuase `WebView` class wraps a native window.

haruhisa-shin replied:
Indeed. I will check.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -42,22 +44,41 @@ AutomationSessionClient::AutomationSessionClient(const String& sessionIdentifier
-    }
-    if (!found)
-        completionHandler(nullptr);
+    m_webViews.clear();
```
m_webViews will be automatically destroyed. Why do you have to call `clear()` here?

haruhisa-shin replied:
As you say, this `clear()` is unnecessary. Deleted.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.cpp:
```diff
@@ -73,6 +94,18 @@ void AutomationSessionClient::didDisconnectFromRemote(WebKit::WebAutomationSessi
+void AutomationSessionClient::releaseWebView(HWND hwnd)
+{
+    if (auto webView = m_webViews.take(hwnd))
+        webView->close();
```
Do you need to call `close()` here? What happens if you just release  the lifetime of webview?

fujii replied:
Never mind. I get the idea.

fujii commented on Source/WebKit/UIProcess/win/AutomationClientWin.h:
```diff
@@ -40,14 +42,21 @@ namespace WebKit {
     String m_sessionIdentifier;
+
+    static void close(WKPageRef, const void*);
+    HashMap<HWND, RefPtr<WebView>> m_webViews;
```
We no longer have the reason to use HWND here. We can use WebPageProxy* or NativeWindowHandle.

fujii replied:
I think we don't need HashMap anymore. We can do `Vector<Ref<WebView>> m_webViews`  or `HashSet<Ref<WebView>>` here.

fujii replied:
You can use `WebView::page()` to get a `WebPageProxy*` for matching.

haruhisa-shin replied:
Thank you for pointing this out.
The `HWND` has been removed as it is no longer needed.
WebViews management has been changed to `HashSet`.

haruhisa-shin committed:
[WebDriver][win] WebDriver launches MiniBrowser for automation
https://bugs.webkit.org/show_bug.cgi?id=282972

Reviewed by Fujii Hironori.

The win port WebDriver supports connecting to a browser
that is already running.
However, win port WebDriver has not yet implemented
the feature of launching the browser by itself.
This makes it difficult to run the WebDriverTests.

This change adds a feature to the WebDriver to launch
a browser to connect if target is not specified.

Before this change, the --target(-t) option is
mandatory for win port WebDriver.

This change allows the --target option to be omitted.
In that case, the WebDriver will scan freeport and
launching a MiniBrowser that listens on 127.0.0.1:(freeport).
This MiniBrowser is hidden, but WebDriver can connect to
and automate it.

Canonical link: https://commits.webkit.org/287053@main
